kv/kvnemesis: TestKVNemesisMultiNode failed

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9661034?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9661034?buildTab=artifacts#/) on master @ [1cd507a7c6582fd91bb56123e62bd4fde45c4d22](https://github.com/cockroachdb/cockroach/commits/1cd507a7c6582fd91bb56123e62bd4fde45c4d22):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1865204116
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 7605291541804074861
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2082477957
    kvnemesis.go:165: range 65 (/Table/100/"00532edf314db371") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1681814839621095004 IntentAge:0 GCBytesAge:13445 LiveBytes:212 LiveCount:4 KeyBytes:435 KeyCount:11 ValBytes:290 ValCount:17 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:1 RangeKeyBytes:51 RangeValCount:1 RangeValBytes:10 SysBytes:10401 SysCount:20 AbortSpanBytes:1199}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1681814839621095004 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:-6 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:165: range 78 (/Table/100/"0d1b2742d6b6062f") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1681814839812403396 IntentAge:0 GCBytesAge:41393 LiveBytes:53 LiveCount:1 KeyBytes:750 KeyCount:14 ValBytes:583 ValCount:38 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:1 RangeKeyBytes:51 RangeValCount:1 RangeValBytes:10 SysBytes:2488 SysCount:21 AbortSpanBytes:1199}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1681814839812403396 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:6 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2082477957/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2082477957/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2082477957/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2082477957/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 2
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1865204116
--- FAIL: TestKVNemesisMultiNode (70.36s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27104


Epic CRDB-27234

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9682828?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9682828?buildTab=artifacts#/) on master @ [a4ab6b6243f509e62993b4d81855f23470d9901a](https://github.com/cockroachdb/cockroach/commits/a4ab6b6243f509e62993b4d81855f23470d9901a):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1884782417
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 7148618387478836565
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1659314035
    kvnemesis.go:165: range 75 (/Table/100/"bb01c544f45f5f14") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1681899478668119526 IntentAge:0 GCBytesAge:73101 LiveBytes:53 LiveCount:1 KeyBytes:693 KeyCount:13 ValBytes:559 ValCount:35 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:3 RangeKeyBytes:171 RangeValCount:5 RangeValBytes:51 SysBytes:4683 SysCount:18 AbortSpanBytes:941}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1681899478668119526 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:6 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1659314035/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1659314035/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1659314035/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1659314035/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1884782417
--- FAIL: TestKVNemesisMultiNode (109.79s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9740681?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9740681?buildTab=artifacts#/) on master @ [10665f9856d9c7ca51557bdf5b48fe9df796228b](https://github.com/cockroachdb/cockroach/commits/10665f9856d9c7ca51557bdf5b48fe9df796228b):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1227749579
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 7800460224026642564
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3322945106
    kvnemesis.go:165: range 73 (/Table/100/"b155b47704015b5a") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1682157779371627510 IntentAge:0 GCBytesAge:76056 LiveBytes:53 LiveCount:1 KeyBytes:1173 KeyCount:25 ValBytes:839 ValCount:54 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:4 RangeKeyBytes:222 RangeValCount:6 RangeValBytes:62 SysBytes:3593 SysCount:14 AbortSpanBytes:598}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1682157779371627510 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:18 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:165: range 106 (/Table/100/"557a2358987d8669") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1682157779926807589 IntentAge:0 GCBytesAge:46609 LiveBytes:53 LiveCount:1 KeyBytes:519 KeyCount:11 ValBytes:384 ValCount:24 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:2 RangeKeyBytes:137 RangeValCount:5 RangeValBytes:51 SysBytes:1076 SysCount:14 AbortSpanBytes:683}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1682157779926807589 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:-18 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3322945106/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3322945106/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3322945106/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3322945106/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 2
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1227749579
--- FAIL: TestKVNemesisMultiNode (119.19s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9753033?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9753033?buildTab=artifacts#/) on master @ [1f3419e178bdba544f74d9c9e14a4682efd18028](https://github.com/cockroachdb/cockroach/commits/1f3419e178bdba544f74d9c9e14a4682efd18028):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1377543860
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 1024415409345674575
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis752674863
    kvnemesis.go:165: range 115 (/Table/100/"4f372447fddf9089") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1682248667918326218 IntentAge:0 GCBytesAge:39780 LiveBytes:0 LiveCount:0 KeyBytes:624 KeyCount:12 ValBytes:490 ValCount:31 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:1 RangeKeyBytes:51 RangeValCount:1 RangeValBytes:10 SysBytes:3401 SysCount:21 AbortSpanBytes:1293}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1682248667918326218 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:-6 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis752674863/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis752674863/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis752674863/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis752674863/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1377543860
--- FAIL: TestKVNemesisMultiNode (96.03s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9763110?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9763110?buildTab=artifacts#/) on master @ [1f3419e178bdba544f74d9c9e14a4682efd18028](https://github.com/cockroachdb/cockroach/commits/1f3419e178bdba544f74d9c9e14a4682efd18028):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode4065196892
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 7646607177314613521
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3681986409
    kvnemesis.go:165: range 147 (/Table/100/"7bc4d1e4075c7520") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1682329766680715445 IntentAge:0 GCBytesAge:34428 LiveBytes:0 LiveCount:0 KeyBytes:285 KeyCount:5 ValBytes:243 ValCount:15 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:1169 SysCount:10 AbortSpanBytes:420}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1682329766680715445 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:6 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3681986409/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3681986409/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3681986409/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3681986409/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode4065196892
--- FAIL: TestKVNemesisMultiNode (172.17s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9806285?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9806285?buildTab=artifacts#/) on master @ [9fb03ee9b365d656cc26728cc87f682a385d511b](https://github.com/cockroachdb/cockroach/commits/9fb03ee9b365d656cc26728cc87f682a385d511b):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode2090673101
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 753312517745099591
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1232282953
    kvnemesis.go:165: error applying x.AdminMerge(ctx, tk(6314721792752128086)) // kv/kvserver/replica_command.go:808: merge failed: waiting for all right-hand replicas to initialize: operation "wait for replicas init" timed out after 5.002s (given timeout 5s): grpc: context deadline exceeded [code 4/DeadlineExceeded]: kv/kvserver/replica_command.go:808: merge failed: waiting for all right-hand replicas to initialize: operation "wait for replicas init" timed out after 5.002s (given timeout 5s): grpc: context deadline exceeded [code 4/DeadlineExceeded]
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1232282953/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1232282953/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1232282953/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1232282953/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode2090673101
--- FAIL: TestKVNemesisMultiNode (19.84s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9918978?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9918978?buildTab=artifacts#/) on master @ [9e0b06bf61700d5192f171c6ebaec7e8a10fa6e6](https://github.com/cockroachdb/cockroach/commits/9e0b06bf61700d5192f171c6ebaec7e8a10fa6e6):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode2226663341
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 8827703835256234507
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis190548753
    kvnemesis.go:165: range 66 (/Table/100/"43181f1ddecdf899") RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1683110290381382106 IntentAge:0 GCBytesAge:12442 LiveBytes:0 LiveCount:0 KeyBytes:1281 KeyCount:25 ValBytes:975 ValCount:63 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:2 RangeKeyBytes:115 RangeValCount:3 RangeValBytes:31 SysBytes:4629 SysCount:24 AbortSpanBytes:1532}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1683110290381382106 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:-10 SysCount:0 AbortSpanBytes:0}
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis190548753/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis190548753/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis190548753/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis190548753/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode2226663341
--- FAIL: TestKVNemesisMultiNode (16.69s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #102411 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv X-duplicate branch-release-23.1.0]
- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10123808?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10123808?buildTab=artifacts#/) on master @ [69875da00208dd792cb92aee54811e8a660773a9](https://github.com/cockroachdb/cockroach/commits/69875da00208dd792cb92aee54811e8a660773a9):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode2895614279
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 2840821478841235807
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1782552098
    kvnemesis.go:165: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.Snapshot)
          txn.ReverseScan(ctx, tk(2472807650999776716), tk(15928832316702100637), 0) // kv/kvclient/kvcoord/txn_coord_sender.go:754: unexpected txn state: "unnamed" meta={id=40ecf13e key=/Table/100/"08ce982a34f906c6" iso=Snapshot pri=0.02441066 epo=1 ts=1684310680.015681917,0 min=1684310679.797745821,0 seq=0} lock=true stat=STAGING rts=1684310680.015681917,0 wto=false gul=1684310679.807745821,0; heartbeat observed status: PENDING
          txn.Get(ctx, tk(9210135055451966804)) // omitted
          txn.Put(ctx, tk(634611904550602438), sv(495)) // omitted
          b := &kv.Batch{}
          b.Put(tk(3897933467779442998), sv(496)) // TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "unnamed" meta={id=40ecf13e key=/Table/100/"08ce982a34f906c6" iso=Snapshot pri=0.02441066 epo=0 ts=1684310679.905093068,1 min=1684310679.797745821,0 seq=3} lock=true stat=STAGING rts=1684310679.807745821,0 wto=false gul=1684310679.807745821,0 ifw=2
          txn.CommitInBatch(ctx, b) // TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "unnamed" meta={id=40ecf13e key=/Table/100/"08ce982a34f906c6" iso=Snapshot pri=0.02441066 epo=0 ts=1684310679.905093068,1 min=1684310679.797745821,0 seq=3} lock=true stat=STAGING rts=1684310679.807745821,0 wto=false gul=1684310679.807745821,0 ifw=2
          return nil
        }) // kv/kvclient/kvcoord/txn_coord_sender.go:754: unexpected txn state: "unnamed" meta={id=40ecf13e key=/Table/100/"08ce982a34f906c6" iso=Snapshot pri=0.02441066 epo=1 ts=1684310680.015681917,0 min=1684310679.797745821,0 seq=0} lock=true stat=STAGING rts=1684310680.015681917,0 wto=false gul=1684310679.807745821,0; heartbeat observed status: PENDING: kv/kvclient/kvcoord/txn_coord_sender.go:754: unexpected txn state: "unnamed" meta={id=40ecf13e key=/Table/100/"08ce982a34f906c6" iso=Snapshot pri=0.02441066 epo=1 ts=1684310680.015681917,0 min=1684310679.797745821,0 seq=0} lock=true stat=STAGING rts=1684310680.015681917,0 wto=false gul=1684310679.807745821,0; heartbeat observed status: PENDING
    kvnemesis.go:165: error applying x.ReverseScan(ctx, tk(2472807650999776716), tk(15928832316702100637), 0) // kv/kvclient/kvcoord/txn_coord_sender.go:754: unexpected txn state: "unnamed" meta={id=40ecf13e key=/Table/100/"08ce982a34f906c6" iso=Snapshot pri=0.02441066 epo=1 ts=1684310680.015681917,0 min=1684310679.797745821,0 seq=0} lock=true stat=STAGING rts=1684310680.015681917,0 wto=false gul=1684310679.807745821,0; heartbeat observed status: PENDING: kv/kvclient/kvcoord/txn_coord_sender.go:754: unexpected txn state: "unnamed" meta={id=40ecf13e key=/Table/100/"08ce982a34f906c6" iso=Snapshot pri=0.02441066 epo=1 ts=1684310680.015681917,0 min=1684310679.797745821,0 seq=0} lock=true stat=STAGING rts=1684310680.015681917,0 wto=false gul=1684310679.807745821,0; heartbeat observed status: PENDING
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1782552098/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1782552098/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1782552098/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1782552098/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
        	Error:      	Should be zero, but was 2
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode2895614279
--- FAIL: TestKVNemesisMultiNode (28.67s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #102411 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv X-duplicate todo-deprecate.branch-release-23.1.0]
- #101048 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot T-kv branch-release-22.2 no-test-failure-activity]
- #100417 kv/kvnemesis: TestKVNemesisMultiNode failed [C-bug C-test-failure O-robot T-kv branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Unfortunately we don't have surviving artifacts for `kvnemesis.go:165: error applying x.AdminMerge(ctx, tk(6314721792752128086)) // kv/kvserver/replica_command.go:808: merge failed: waiting for all right-hand replicas to initialize: operation "wait for replicas init" timed out after 5.002s (given timeout 5s): grpc: context deadline exceeded [code 4/DeadlineExceeded]: kv/kvserver/replica_command.go:808: merge failed: waiting for all right-hand replicas to initialize: operation "wait for replicas init" timed out after 5.002s (given timeout 5s): grpc: context deadline exceeded [code 4/DeadlineExceeded]`
I'll try to run repro with same seed, but not sure it'll be easy with the size of the test.

We havn't seen many failures after stats were fixed. I managed to repro transaction push timeout with 50k reruns while testing stats:

```
--- FAIL: TestKVNemesisMultiNode (13.16s)
    test_log_scope.go:161: test logs captured to: /tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode3315994629
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 910877016883269571
    kvnemesis_test.go:124: kvnemesis logging to /tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1584584814
    kvnemesis.go:165: error applying x.Scan(ctx, tk(4919960733718109041), tk(8091360691588176074), 0) // failed to push meta={id=d01cdc27 key=/Table/100/"480aacd5586decfa" iso=Snapshot pri=0.06480878 epo=1 ts=1684954142.358901691,1 min=1684954142.358901691,0 seq=2} lock=true stat=PENDING rts=0,0 wto=false gul=0,0: {<nil> 0 {0xc003a2caf8} index:0 1684954142.383387488,0}
    kvnemesis.go:185: failures(verbose): /tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1584584814/failures
        repro steps: /tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1584584814/repro.go
        rangefeed KVs: /tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1584584814/kvs-rangefeed.txt
        scan KVs: /tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1584584814/kvs-scan.txt
    kvnemesis_test.go:207:
                Error Trace:    github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
                                                        github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:160
                Error:          Should be zero, but was 1
                Test:           TestKVNemesisMultiNode
                Messages:       kvnemesis detected failures
    panic.go:522: -- test log scope end --
FAIL
I230524 18:49:03.146045 1 (gostd) testmain.go:100  [-] 1  Test //pkg/kv/kvnemesis:kvnemesis_test exited with error code 1
```

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11123306?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11123306?buildTab=artifacts#/) on master @ [e57e9742146b7fac7f7530fa6224e11121ada92f](https://github.com/cockroachdb/cockroach/commits/e57e9742146b7fac7f7530fa6224e11121ada92f):

Fatal error:

```
panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 7767151 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x1f8006a?, {0x69690a8, 0xc012f9d3b0})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x4b26d80, 0xc0087e0730})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x69690a8?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x12e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:225
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaDecoder).createTracingSpans(0xc0066c7658, {0x69690a8, 0xc0029e5980})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_application_decoder.go:154 +0x4f0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaDecoder).DecodeAndBind(0xc0066c7200?, {0x69690a8, 0xc0029e5980}, {0xc0153d2750?, 0x2?, 0x10?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_application_decoder.go:63 +0x66
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).Decode(...)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:142
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:878 +0x73a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:728 +0x1b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc006662000, 0x2?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:689 +0x14b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc00de44540, {0x69690a8, 0xc012f9d3b0}, {0x694eee0, 0xc006662000}, 0xc006581500)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:416 +0x274
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x69690a8?, 0xc012f9d3b0?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode2457498475
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:283: seed: 8481106198570085994
    test_server_shim.go:98: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
    kvnemesis_test.go:197: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis4034049211
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvnemesis/validator.go
+++ b/pkg/kv/kvnemesis/validator.go
@@ -1232,16 +1232,6 @@ func (v *validator) failIfError(
 ) (ambiguous, hasError bool) {
 	exceptions = append(exceptions[:len(exceptions):len(exceptions)], func(err error) bool {
 		return errors.Is(err, errInjected)
-	}, func(err error) bool {
-		// Work-around for [1].
-		//
-		// TODO(arul): find out why we (as of [2]) sometimes leaking
-		// *TransactionPushError (wrapped in `UnhandledRetryableError`) from
-		// `db.Get`, `db.Scan`, etc.
-		//
-		// [1]: https://github.com/cockroachdb/cockroach/issues/105330
-		// [2]: https://github.com/cockroachdb/cockroach/pull/97779
-		return errors.HasType(err, (*kvpb.UnhandledRetryableError)(nil))
 	})
 	switch r.Type {
 	case ResultType_Unknown:

--- a/pkg/kv/kvserver/batcheval/cmd_push_txn.go
+++ b/pkg/kv/kvserver/batcheval/cmd_push_txn.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/must"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/redact"
 )
@@ -213,38 +214,9 @@ func PushTxn(
 	}
 	reply.PusheeTxn.UpgradePriority(args.PusheeTxn.Priority)
 
-	// If the pusher is aware that the pushee's currently recorded attempt at a
-	// parallel commit failed, either because it found intents at a higher
-	// timestamp than the parallel commit attempt or because it found intents at
-	// a higher epoch than the parallel commit attempt, it should not consider
-	// the pushee to be performing a parallel commit. Its commit status is not
-	// indeterminate.
-	if (knownHigherTimestamp || knownHigherEpoch) && reply.PusheeTxn.Status == roachpb.STAGING {
-		reply.PusheeTxn.Status = roachpb.PENDING
-		reply.PusheeTxn.InFlightWrites = nil
-		// If the pusher is aware that the pushee's currently recorded attempt
-		// at a parallel commit failed, upgrade PUSH_TIMESTAMPs to PUSH_ABORTs.
-		// We don't want to move the transaction back to PENDING, as this is not
-		// (currently) allowed by the recovery protocol. We also don't want to
-		// move the transaction to a new timestamp while retaining the STAGING
-		// status, as this could allow the transaction to enter an implicit
-		// commit state without its knowledge, leading to atomicity violations.
-		//
-		// This has no effect on pushes that fail with a TransactionPushError.
-		// Such pushes will still wait on the pushee to retry its commit and
-		// eventually commit or abort. It also has no effect on expired pushees,
-		// as they would have been aborted anyway. This only impacts pushes
-		// which would have succeeded due to priority mismatches. In these
-		// cases, the push acts the same as a short-circuited transaction
-		// recovery process, because the transaction recovery procedure always
-		// finalizes target transactions, even if initiated by a PUSH_TIMESTAMP.
-		if pushType == kvpb.PUSH_TIMESTAMP {
-			pushType = kvpb.PUSH_ABORT
-		}
-	}
-
 	pusherIso, pusheeIso := args.PusherTxn.IsoLevel, reply.PusheeTxn.IsoLevel
 	pusherPri, pusheePri := args.PusherTxn.Priority, reply.PusheeTxn.Priority
+	pusheeStatus := reply.PusheeTxn.Status
 	var pusherWins bool
 	var reason string
 	switch {
@@ -258,7 +230,7 @@ func PushTxn(
 		// If just attempting to cleanup old or already-committed txns,
 		// pusher always fails.
 		pusherWins = false
-	case txnwait.CanPushWithPriority(pushType, pusherIso, pusheeIso, pusherPri, pusheePri):
+	case txnwait.CanPushWithPriority(pushType, pusherIso, pusheeIso, pusherPri, pusheePri, pusheeStatus):
 		reason = "pusher has priority"
 		pusherWins = true
 	case args.Force:
@@ -282,11 +254,40 @@ func PushTxn(
 	// If the pushed transaction is in the staging state, we can't change its
 	// record without first going through the transaction recovery process and
 	// attempting to finalize it.
+	pusheeStaging := pusheeStatus == roachpb.STAGING
+	// However, if the pusher is aware that the pushee's currently recorded
+	// attempt at a parallel commit failed, either because it found intents at a
+	// higher timestamp than the parallel commit attempt or because it found
+	// intents at a higher epoch than the parallel commit attempt, it should not
+	// consider the pushee to be performing a parallel commit. Its commit status
+	// is not indeterminate.
+	pusheeStagingFailed := pusheeStaging && (knownHigherTimestamp || knownHigherEpoch)
 	recoverOnFailedPush := cArgs.EvalCtx.EvalKnobs().RecoverIndeterminateCommitsOnFailedPushes
-	if reply.PusheeTxn.Status == roachpb.STAGING && (pusherWins || recoverOnFailedPush) {
+	if pusheeStaging && !pusheeStagingFailed && (pusherWins || recoverOnFailedPush) {
 		err := kvpb.NewIndeterminateCommitError(reply.PusheeTxn)
 		log.VEventf(ctx, 1, "%v", err)
 		return result.Result{}, err
+	}
+
+	// If the pusher is aware that the pushee's currently recorded attempt at a
+	// parallel commit failed, upgrade PUSH_TIMESTAMPs to PUSH_ABORTs. We don't
+	// want to move the transaction back to PENDING, as this is not (currently)
+	// allowed by the recovery protocol. We also don't want to move the
+	// transaction to a new timestamp while retaining the STAGING status, as this
+	// could allow the transaction to enter an implicit commit state without its
+	// knowledge, leading to atomicity violations.
+	//
+	// This has no effect on pushes that fail with a TransactionPushError. Such
+	// pushes will still wait on the pushee to retry its commit and eventually
+	// commit or abort. It also has no effect on expired pushees, as they would
+	// have been aborted anyway. This only impacts pushes which would have
+	// succeeded due to priority mismatches. In these cases, the push acts the
+	// same as a short-circuited transaction recovery process, because the
+	// transaction recovery procedure always finalizes target transactions, even
+	// if initiated by a PUSH_TIMESTAMP.
+	if pusheeStaging && pusherWins && pushType == kvpb.PUSH_TIMESTAMP {
+		_ = must.True(ctx, pusheeStagingFailed, "parallel commit must be known to have failed for push to succeed")
+		pushType = kvpb.PUSH_ABORT
 	}
 
 	if !pusherWins {
@@ -306,6 +307,8 @@ func PushTxn(
 		// Forward the timestamp to accommodate AbortSpan GC. See method comment for
 		// details.
 		reply.PusheeTxn.WriteTimestamp.Forward(reply.PusheeTxn.LastActive())
+		// If the transaction was previously staging, clear its in-flight writes.
+		reply.PusheeTxn.InFlightWrites = nil
 		// If the transaction record was already present, persist the updates to it.
 		// If not, then we don't want to create it. This could allow for finalized
 		// transactions to be revived. Instead, we obey the invariant that only the

--- a/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
+++ b/pkg/kv/kvserver/concurrency/concurrency_manager_test.go
@@ -747,10 +747,11 @@ func (c *cluster) PushTransaction(
 		pusheeTxn, pusheeRecordSig := pusheeRecord.asTxn()
 		pusheeIso := pusheeTxn.IsoLevel
 		pusheePri := pusheeTxn.Priority
+		pusheeStatus := pusheeTxn.Status
 		// NOTE: this logic is adapted from cmd_push_txn.go.
 		var pusherWins bool
 		switch {
-		case pusheeTxn.Status.IsFinalized():
+		case pusheeStatus.IsFinalized():
 			// Already finalized.
 			return pusheeTxn, nil
 		case pushType == kvpb.PUSH_TIMESTAMP && pushTo.LessEq(pusheeTxn.WriteTimestamp):
@@ -758,7 +759,7 @@ func (c *cluster) PushTransaction(
 			return pusheeTxn, nil
 		case pushType == kvpb.PUSH_TOUCH:
 			pusherWins = false
-		case txnwait.CanPushWithPriority(pushType, pusherIso, pusheeIso, pusherPri, pusheePri):
+		case txnwait.CanPushWithPriority(pushType, pusherIso, pusheeIso, pusherPri, pusheePri, pusheeStatus):
 			pusherWins = true
 		default:
 			pusherWins = false

--- a/pkg/kv/kvserver/concurrency/lock_table_waiter.go
+++ b/pkg/kv/kvserver/concurrency/lock_table_waiter.go
@@ -1298,7 +1298,11 @@ func canPushWithPriority(req Request, s waitingState) bool {
 	}
 	pusheeIso = s.txn.IsoLevel
 	pusheePri = s.txn.Priority
-	return txnwait.CanPushWithPriority(pushType, pusherIso, pusheeIso, pusherPri, pusheePri)
+	// We assume that the pushee is in the PENDING state when deciding whether
+	// to push. A push may determine that the pushee is STAGING or has already
+	// been finalized.
+	pusheeStatus := roachpb.PENDING
+	return txnwait.CanPushWithPriority(pushType, pusherIso, pusheeIso, pusherPri, pusheePri, pusheeStatus)
 }
 
 func logResolveIntent(ctx context.Context, intent roachpb.LockUpdate) {

--- a/pkg/kv/kvserver/replica_send.go
+++ b/pkg/kv/kvserver/replica_send.go
@@ -773,7 +773,7 @@ func (r *Replica) handleTransactionPushError(
 	dontRetry := r.store.cfg.TestingKnobs.DontRetryPushTxnFailures
 	if !dontRetry && ba.IsSinglePushTxnRequest() {
 		pushReq := ba.Requests[0].GetInner().(*kvpb.PushTxnRequest)
-		dontRetry = txnwait.ShouldPushImmediately(pushReq)
+		dontRetry = txnwait.ShouldPushImmediately(pushReq, t.PusheeTxn.Status)
 	}
 	if dontRetry {
 		return g, pErr

--- a/pkg/kv/kvserver/txn_recovery_integration_test.go
+++ b/pkg/kv/kvserver/txn_recovery_integration_test.go
@@ -372,20 +372,6 @@ func TestTxnRecoveryFromStagingWithoutHighPriority(t *testing.T) {
 			pErrC <- pErr
 		}))
 
-		// If the pushee is not serializable and the pusher is reading, we
-		// currently expect the conflicting operation to immediately succeed,
-		// sometimes with an error, sometimes not. See #105330.
-		if pusheeIsoLevel.ToleratesWriteSkew() && !pusherWriting {
-			pErr = <-pErrC
-			if pusheeCommits {
-				require.Nil(t, pErr, "error: %s", pErr)
-			} else {
-				require.NotNil(t, pErr)
-				require.Regexp(t, "failed to push", pErr)
-			}
-			return
-		}
-
 		// Wait for the conflict to push and be queued in the txn wait queue.
 		testutils.SucceedsSoon(t, func() error {
 			select {

--- a/pkg/kv/kvserver/txn_recovery_integration_test.go
+++ b/pkg/kv/kvserver/txn_recovery_integration_test.go
@@ -18,6 +18,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/isolation"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -25,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
 
@@ -286,6 +288,141 @@ func TestTxnRecoveryFromStagingWithHighPriority(t *testing.T) {
 			})
 		})
 	})
+}
+
+// TestTxnRecoveryFromStagingWithoutHighPriority tests that the transaction
+// recovery process is NOT initiated by a normal-priority operation which
+// encounters a staging transaction. Instead, the normal-priority operation
+// waits for the committing transaction to complete. The test contains a subtest
+// for each of the combinations of the following options:
+//
+//   - pusheeIsoLevel: configures the isolation level of the pushee (committing)
+//     transaction. Isolation levels affect the behavior of pushes of pending
+//     transactions, but not of staging transactions.
+//
+//   - pusheeCommits: configures whether or not the staging transaction is
+//     implicitly and, eventually, explicitly committed or not.
+//
+//   - pusherWriting: configures whether or not the conflicting operation is a
+//     read (false) or a write (true), which dictates the kind of push operation
+//     dispatched against the staging transaction.
+func TestTxnRecoveryFromStagingWithoutHighPriority(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	run := func(t *testing.T, pusheeIsoLevel isolation.Level, pusheeCommits, pusherWriting bool) {
+		stopper := stop.NewStopper()
+		defer stopper.Stop(ctx)
+		manual := timeutil.NewManualTime(timeutil.Unix(0, 123))
+		cfg := TestStoreConfig(hlc.NewClockForTesting(manual))
+		store := createTestStoreWithConfig(ctx, t, stopper, testStoreOpts{createSystemRanges: true}, &cfg)
+
+		// Create a transaction that will get stuck performing a parallel
+		// commit.
+		keyA, keyB := roachpb.Key("a"), roachpb.Key("b")
+		txn := newTransaction("txn", keyA, 1, store.Clock())
+		txn.IsoLevel = pusheeIsoLevel
+
+		// Issue two writes, which will be considered in-flight at the time of
+		// the transaction's EndTxn request.
+		keyAVal := []byte("value")
+		pArgs := putArgs(keyA, keyAVal)
+		pArgs.Sequence = 1
+		h := kvpb.Header{Txn: txn}
+		_, pErr := kv.SendWrappedWith(ctx, store.TestSender(), h, &pArgs)
+		require.Nil(t, pErr, "error: %s", pErr)
+
+		pArgs = putArgs(keyB, []byte("value2"))
+		pArgs.Sequence = 2
+		h2 := kvpb.Header{Txn: txn.Clone()}
+		if !pusheeCommits {
+			// If we're not going to have the pushee commit, make sure it never enters
+			// the implicit commit state by bumping the timestamp of one of its writes.
+			manual.Advance(100)
+			h2.Txn.WriteTimestamp = store.Clock().Now()
+		}
+		_, pErr = kv.SendWrappedWith(ctx, store.TestSender(), h2, &pArgs)
+		require.Nil(t, pErr, "error: %s", pErr)
+
+		// Issue a parallel commit, which will put the transaction into a
+		// STAGING state. Include both writes as the EndTxn's in-flight writes.
+		et, etH := endTxnArgs(txn, true)
+		et.InFlightWrites = []roachpb.SequencedWrite{
+			{Key: keyA, Sequence: 1},
+			{Key: keyB, Sequence: 2},
+		}
+		etReply, pErr := kv.SendWrappedWith(ctx, store.TestSender(), etH, &et)
+		require.Nil(t, pErr, "error: %s", pErr)
+		require.Equal(t, roachpb.STAGING, etReply.Header().Txn.Status)
+
+		// Issue a conflicting, normal-priority operation.
+		var conflictArgs kvpb.Request
+		if pusherWriting {
+			pArgs = putArgs(keyB, []byte("value3"))
+			conflictArgs = &pArgs
+		} else {
+			gArgs := getArgs(keyB)
+			conflictArgs = &gArgs
+		}
+		manual.Advance(100)
+		pErrC := make(chan *kvpb.Error, 1)
+		require.NoError(t, stopper.RunAsyncTask(ctx, "conflict", func(ctx context.Context) {
+			_, pErr := kv.SendWrapped(ctx, store.TestSender(), conflictArgs)
+			pErrC <- pErr
+		}))
+
+		// If the pushee is not serializable and the pusher is reading, we
+		// currently expect the conflicting operation to immediately succeed,
+		// sometimes with an error, sometimes not. See #105330.
+		if pusheeIsoLevel.ToleratesWriteSkew() && !pusherWriting {
+			pErr = <-pErrC
+			if pusheeCommits {
+				require.Nil(t, pErr, "error: %s", pErr)
+			} else {
+				require.NotNil(t, pErr)
+				require.Regexp(t, "failed to push", pErr)
+			}
+			return
+		}
+
+		// Wait for the conflict to push and be queued in the txn wait queue.
+		testutils.SucceedsSoon(t, func() error {
+			select {
+			case pErr := <-pErrC:
+				t.Fatalf("conflicting operation unexpectedly completed: pErr=%s", pErr)
+			default:
+			}
+			if v := store.txnWaitMetrics.PusherWaiting.Value(); v != 1 {
+				return errors.Errorf("expected 1 pusher waiting, found %d", v)
+			}
+			return nil
+		})
+
+		// Finalize the STAGING txn, either by committing it or by aborting it.
+		et2, et2H := endTxnArgs(txn, pusheeCommits)
+		etReply, pErr = kv.SendWrappedWith(ctx, store.TestSender(), et2H, &et2)
+		require.Nil(t, pErr, "error: %s", pErr)
+		expStatus := roachpb.COMMITTED
+		if !pusheeCommits {
+			expStatus = roachpb.ABORTED
+		}
+		require.Equal(t, expStatus, etReply.Header().Txn.Status)
+
+		// This will unblock the conflicting operation, which should succeed.
+		pErr = <-pErrC
+		require.Nil(t, pErr, "error: %s", pErr)
+	}
+
+	for _, pusheeIsoLevel := range isolation.Levels() {
+		t.Run("pushee_iso_level="+pusheeIsoLevel.String(), func(t *testing.T) {
+			testutils.RunTrueAndFalse(t, "pushee_commits", func(t *testing.T, pusheeCommits bool) {
+				testutils.RunTrueAndFalse(t, "pusher_writing", func(t *testing.T, pusherWriting bool) {
+					run(t, pusheeIsoLevel, pusheeCommits, pusherWriting)
+				})
+			})
+		})
+	}
 }
 
 // TestTxnClearRangeIntents tests whether a ClearRange call blindly removes

--- a/pkg/kv/kvserver/txnwait/queue.go
+++ b/pkg/kv/kvserver/txnwait/queue.go
@@ -70,7 +70,7 @@ func TestingOverrideTxnLivenessThreshold(t time.Duration) func() {
 // proceed without queueing. This is true for pushes which are neither
 // ABORT nor TIMESTAMP, but also for ABORT and TIMESTAMP pushes where
 // the pushee has min priority or pusher has max priority.
-func ShouldPushImmediately(req *kvpb.PushTxnRequest) bool {
+func ShouldPushImmediately(req *kvpb.PushTxnRequest, pusheeStatus roachpb.TransactionStatus) bool {
 	if req.Force {
 		return true
 	}
@@ -78,15 +78,18 @@ func ShouldPushImmediately(req *kvpb.PushTxnRequest) bool {
 		req.PushType,
 		req.PusherTxn.IsoLevel, req.PusheeTxn.IsoLevel,
 		req.PusherTxn.Priority, req.PusheeTxn.Priority,
+		pusheeStatus,
 	)
 }
 
 // CanPushWithPriority returns true if the pusher can perform the specified push
-// type on the pushee, based on the two txns' isolation levels and priorities.
+// type on the pushee, based on the two txns' isolation levels, their priorities,
+// and the pushee's status.
 func CanPushWithPriority(
 	pushType kvpb.PushTxnType,
 	pusherIso, pusheeIso isolation.Level,
 	pusherPri, pusheePri enginepb.TxnPriority,
+	pusheeStatus roachpb.TransactionStatus,
 ) bool {
 	// Normalize the transaction priorities so that normal user priorities are
 	// considered equal for the purposes of pushing.
@@ -103,6 +106,15 @@ func CanPushWithPriority(
 	case kvpb.PUSH_ABORT:
 		return pusherPri > pusheePri
 	case kvpb.PUSH_TIMESTAMP:
+		// If the pushee transaction is STAGING, only let the PUSH_TIMESTAMP through
+		// to disrupt the transaction commit if the pusher has a higher priority. If
+		// the priorities are equal, the PUSH_TIMESTAMP should wait for the commit
+		// to complete.
+		if pusheeStatus == roachpb.STAGING {
+			return pusherPri > pusheePri
+		}
+		// Otherwise, the pushee has not yet started committing...
+
 		// If the pushee transaction tolerates write skew, the PUSH_TIMESTAMP is
 		// harmless, so let it through.
 		return pusheeIso.ToleratesWriteSkew() ||
@@ -476,10 +488,6 @@ func (q *Queue) releaseWaitingQueriesLocked(ctx context.Context, txnID uuid.UUID
 func (q *Queue) MaybeWaitForPush(
 	ctx context.Context, req *kvpb.PushTxnRequest,
 ) (*kvpb.PushTxnResponse, *kvpb.Error) {
-	if ShouldPushImmediately(req) {
-		return nil, nil
-	}
-
 	q.mu.Lock()
 	// If the txn wait queue is not enabled or if the request is not
 	// contained within the replica, do nothing. The request can fall
@@ -501,6 +509,9 @@ func (q *Queue) MaybeWaitForPush(
 	if txn := pending.getTxn(); isPushed(req, txn) {
 		q.mu.Unlock()
 		return createPushTxnResponse(txn), nil
+	} else if ShouldPushImmediately(req, txn.Status) {
+		q.mu.Unlock()
+		return nil, nil
 	}
 
 	push := &waitingPush{

--- a/pkg/kv/kvserver/txnwait/queue_test.go
+++ b/pkg/kv/kvserver/txnwait/queue_test.go
@@ -161,7 +161,8 @@ func TestShouldPushImmediately(t *testing.T) {
 					Priority: test.pusheePri,
 				},
 			}
-			shouldPush := ShouldPushImmediately(&req)
+			pusheeStatus := roachpb.PENDING
+			shouldPush := ShouldPushImmediately(&req, pusheeStatus)
 			require.Equal(t, test.shouldPush, shouldPush)
 		})
 	}
@@ -179,6 +180,7 @@ func testCanPushWithPriorityPushAbort(t *testing.T) {
 	max := enginepb.MaxTxnPriority
 	mid1 := enginepb.TxnPriority(1)
 	mid2 := enginepb.TxnPriority(2)
+	statuses := []roachpb.TransactionStatus{roachpb.PENDING, roachpb.STAGING}
 	testCases := []struct {
 		pusherPri enginepb.TxnPriority
 		pusheePri enginepb.TxnPriority
@@ -201,22 +203,31 @@ func testCanPushWithPriorityPushAbort(t *testing.T) {
 		{max, mid2, true},
 		{max, max, false},
 	}
-	// NOTE: the behavior of PUSH_ABORT pushes is agnostic to isolation levels.
-	for _, pusherIso := range isolation.Levels() {
-		for _, pusheeIso := range isolation.Levels() {
-			for _, test := range testCases {
-				name := fmt.Sprintf("pusherIso=%s/pusheeIso=%s/pusherPri=%d/pusheePri=%d",
-					pusherIso, pusheeIso, test.pusherPri, test.pusheePri)
-				t.Run(name, func(t *testing.T) {
-					canPush := CanPushWithPriority(kvpb.PUSH_ABORT, pusherIso, pusheeIso, test.pusherPri, test.pusheePri)
-					require.Equal(t, test.exp, canPush)
-				})
+	// NOTE: the behavior of PUSH_ABORT pushes is agnostic to pushee status.
+	for _, pusheeStatus := range statuses {
+		// NOTE: the behavior of PUSH_ABORT pushes is agnostic to isolation levels.
+		for _, pusherIso := range isolation.Levels() {
+			for _, pusheeIso := range isolation.Levels() {
+				for _, test := range testCases {
+					name := fmt.Sprintf("pusheeStatus=%s/pusherIso=%s/pusheeIso=%s/pusherPri=%d/pusheePri=%d",
+						pusheeStatus, pusherIso, pusheeIso, test.pusherPri, test.pusheePri)
+					t.Run(name, func(t *testing.T) {
+						canPush := CanPushWithPriority(
+							kvpb.PUSH_ABORT, pusherIso, pusheeIso, test.pusherPri, test.pusheePri, pusheeStatus)
+						require.Equal(t, test.exp, canPush)
+					})
+				}
 			}
 		}
 	}
 }
 
 func testCanPushWithPriorityPushTimestamp(t *testing.T) {
+	t.Run("pusheeStatus="+roachpb.PENDING.String(), testCanPushWithPriorityPushTimestampPusheePending)
+	t.Run("pusheeStatus="+roachpb.STAGING.String(), testCanPushWithPriorityPushTimestampPusheeStaging)
+}
+
+func testCanPushWithPriorityPushTimestampPusheePending(t *testing.T) {
 	SSI := isolation.Serializable
 	SI := isolation.Snapshot
 	RC := isolation.ReadCommitted
@@ -389,9 +400,54 @@ func testCanPushWithPriorityPushTimestamp(t *testing.T) {
 		name := fmt.Sprintf("pusherIso=%s/pusheeIso=%s/pusherPri=%d/pusheePri=%d",
 			test.pusherIso, test.pusheeIso, test.pusherPri, test.pusheePri)
 		t.Run(name, func(t *testing.T) {
-			canPush := CanPushWithPriority(kvpb.PUSH_TIMESTAMP, test.pusherIso, test.pusheeIso, test.pusherPri, test.pusheePri)
+			canPush := CanPushWithPriority(
+				kvpb.PUSH_TIMESTAMP, test.pusherIso, test.pusheeIso, test.pusherPri, test.pusheePri, roachpb.PENDING)
 			require.Equal(t, test.exp, canPush)
 		})
+	}
+}
+
+func testCanPushWithPriorityPushTimestampPusheeStaging(t *testing.T) {
+	min := enginepb.MinTxnPriority
+	max := enginepb.MaxTxnPriority
+	mid1 := enginepb.TxnPriority(1)
+	mid2 := enginepb.TxnPriority(2)
+	testCases := []struct {
+		pusherPri enginepb.TxnPriority
+		pusheePri enginepb.TxnPriority
+		exp       bool
+	}{
+		{min, min, false},
+		{min, mid1, false},
+		{min, mid2, false},
+		{min, max, false},
+		{mid1, min, true},
+		{mid1, mid1, false},
+		{mid1, mid2, false},
+		{mid1, max, false},
+		{mid2, min, true},
+		{mid2, mid1, false},
+		{mid2, mid2, false},
+		{mid2, max, false},
+		{max, min, true},
+		{max, mid1, true},
+		{max, mid2, true},
+		{max, max, false},
+	}
+	// NOTE: the behavior of PUSH_TIMESTAMP pushes is agnostic to isolation levels
+	// when the pushee transaction is STAGING.
+	for _, pusherIso := range isolation.Levels() {
+		for _, pusheeIso := range isolation.Levels() {
+			for _, test := range testCases {
+				name := fmt.Sprintf("pusherIso=%s/pusheeIso=%s/pusherPri=%d/pusheePri=%d",
+					pusherIso, pusheeIso, test.pusherPri, test.pusheePri)
+				t.Run(name, func(t *testing.T) {
+					canPush := CanPushWithPriority(
+						kvpb.PUSH_TIMESTAMP, pusherIso, pusheeIso, test.pusherPri, test.pusheePri, roachpb.STAGING)
+					require.Equal(t, test.exp, canPush)
+				})
+			}
+		}
 	}
 }
 
@@ -401,18 +457,23 @@ func testCanPushWithPriorityPushTouch(t *testing.T) {
 	mid1 := enginepb.TxnPriority(1)
 	mid2 := enginepb.TxnPriority(2)
 	priorities := []enginepb.TxnPriority{min, mid1, mid2, max}
-	// NOTE: the behavior of PUSH_TOUCH pushes is agnostic to isolation levels.
-	for _, pusherIso := range isolation.Levels() {
-		for _, pusheeIso := range isolation.Levels() {
-			// NOTE: the behavior of PUSH_TOUCH pushes is agnostic to txn priorities.
-			for _, pusherPri := range priorities {
-				for _, pusheePri := range priorities {
-					name := fmt.Sprintf("pusherIso=%s/pusheeIso=%s/pusherPri=%d/pusheePri=%d",
-						pusherIso, pusheeIso, pusherPri, pusheePri)
-					t.Run(name, func(t *testing.T) {
-						canPush := CanPushWithPriority(kvpb.PUSH_TOUCH, pusherIso, pusheeIso, pusherPri, pusheePri)
-						require.True(t, canPush)
-					})
+	statuses := []roachpb.TransactionStatus{roachpb.PENDING, roachpb.STAGING}
+	// NOTE: the behavior of PUSH_TOUCH pushes is agnostic to pushee status.
+	for _, pusheeStatus := range statuses {
+		// NOTE: the behavior of PUSH_TOUCH pushes is agnostic to isolation levels.
+		for _, pusherIso := range isolation.Levels() {
+			for _, pusheeIso := range isolation.Levels() {
+				// NOTE: the behavior of PUSH_TOUCH pushes is agnostic to txn priorities.
+				for _, pusherPri := range priorities {
+					for _, pusheePri := range priorities {
+						name := fmt.Sprintf("pusheeStatus=%s/pusherIso=%s/pusheeIso=%s/pusherPri=%d/pusheePri=%d",
+							pusheeStatus, pusherIso, pusheeIso, pusherPri, pusheePri)
+						t.Run(name, func(t *testing.T) {
+							canPush := CanPushWithPriority(
+								kvpb.PUSH_TOUCH, pusherIso, pusheeIso, pusherPri, pusheePri, pusheeStatus)
+							require.True(t, canPush)
+						})
+					}
 				}
 			}
 		}

